src/main/java/io/knowledgeassets/myskills/server/user/profile/UserProfileService.java

```diff
@@ -0,0 +1,240 @@
+import io.knowledgeassets.myskills.server.user.User;
+import io.knowledgeassets.myskills.server.user.UserRepository;
+import io.knowledgeassets.myskills.server.userskill.UserSkill;
+import io.knowledgeassets.myskills.server.userskill.UserSkillRepository;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.poi.EmptyFileException;
+import org.apache.poi.openxml4j.exceptions.NotOfficeXmlFileException;
+import org.apache.poi.xwpf.usermodel.IBody;
+import org.apache.poi.xwpf.usermodel.XWPFDocument;
+import org.apache.poi.xwpf.usermodel.XWPFParagraph;
+import org.apache.poi.xwpf.usermodel.XWPFRun;
+import org.apache.poi.xwpf.usermodel.XWPFTableCell;
+import org.apache.xmlbeans.XmlCursor;
+import org.openxmlformats.schemas.wordprocessingml.x2006.main.CTPPr;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.core.io.ClassPathResource;
+import org.springframework.core.io.FileSystemResource;
+import org.springframework.stereotype.Service;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.StreamSupport;
+
+import static java.lang.String.format;
+import static java.util.Optional.ofNullable;
+import static java.util.Objects.requireNonNull;
+
+@Service
+@Slf4j
+public class UserProfileService {
+
+	private static final SimpleDateFormat SDF = new SimpleDateFormat("dd.MM.yyyy");
+
+	private static final ClassPathResource DEFAULT_TEMPLATE_RESOURCE = new ClassPathResource("templates/user-profile.docx");
+
+	private final UserRepository userRepository;
+	private final UserSkillRepository userSkillRepository;
+	private final String templatePath;
+
+	public UserProfileService(UserRepository userRepository,
+							  UserSkillRepository userSkillRepository,
+							  @Value("${templates.user.profile.path:#{null}}") String templatePath) {
+		this.userRepository = requireNonNull(userRepository);
+		this.userSkillRepository = requireNonNull(userSkillRepository);
+		this.templatePath = templatePath;
+	}
+
+	/**
+	 * Retrieves MS Word document (DOCX) with anonymous user profile data.
+	 * @param referenceId user reference id
+	 * @return MS Word document with anonymous user profile data
+	 */
+	public InputStream getAnonymousUserProfileDocument(String referenceId) {
+
+		final User user = getUserByReferenceId(referenceId);
+
+		try (final InputStream is = getTemplate();
+			 final XWPFDocument document = new XWPFDocument(is)) {
+			replacePlaceholders(document, user);
+			final ByteArrayOutputStream b = new ByteArrayOutputStream();
+			document.write(b);
+			return new ByteArrayInputStream(b.toByteArray());
```

Comment:
Since you need a `byte[]` in the controller anyway, what is the purpose of returning an `InputStream` here instead of returning `b.toByteArray()` directly?

Reply:
You are right. There is no point to wrap `byte[]` into an input stream. That was mistake caused by the old design I developed at first. I changed the method to return `byte[]`. Done.